### PR TITLE
Fixed Biome Modification API not working for json biome

### PR DESCRIPTION
**Pull request by TelepathicGrunt**
Issue report: https://github.com/FabricMC/fabric/issues/1269

It was what I thought. The set of "modified biome registrykeys" was failing and causing Biome Modification API to not apply changes to JSON biomes. In fact, the whole set is completely unnecessary too as each time the RegistryOps is fired, the biome instances inside are unique each time and have no changes from the other times RegistryOps fired. Thus, yeeting the set fixed the issue during my testing.

To double check this PR, use this datapack
[desert_test.zip](https://github.com/FabricMC/fabric/files/6527047/desert_test.zip)

I added a desert well once per chunk in the test mod to the vanilla desert biome. The very center top of the well shows what chunk spawned it when you show chunk boundaries with F3+G. 

Run the test without the datapack by creating a world. Do /locate desert and teleport to it to see the 1 well per chunk. Then exit that world and re-enter it and go to a new desert. See how it is still 1 well per chunk. Repeat these two tests but with the datapack on each time and the wells now still generate once per chunk.

### CODE REVIEW

File: fabric-biome-api-v1/src/testmod/java/net/fabricmc/fabric/test/biome/FabricBiomeTest.java [hunk 1]
```diff
@@ -30,9 +30,7 @@
 import net.minecraft.world.biome.GenerationSettings;
 import net.minecraft.world.biome.SpawnSettings;
 import net.minecraft.world.gen.GenerationStep;
-import net.minecraft.world.gen.feature.ConfiguredFeatures;
-import net.minecraft.world.gen.feature.ConfiguredStructureFeatures;
-import net.minecraft.world.gen.feature.DefaultBiomeFeatures;
+import net.minecraft.world.gen.feature.*;
```
Comment (liach) [inline fabric-biome-api-v1/src/testmod/java/net/fabricmc/fabric/test/biome/FabricBiomeTest.java]: mind use individual imports? and possbly add an actual data pack biome to the test

**Comment by modmuss50** [comment]
Awesome 👍 I think I assumed the modified biomes was for the biome instance and not the reg key...

Idk if a better soultion would be to add a boolean to the biome object to note if its been modified? However it seems that check might not have even been needed in the first place?

**Comment by TelepathicGrunt** [comment]
New problem, an edge case was missed. ~~When there's no json biomes, every biome gets remade as a new instance every time the RegistryOps fires.~~ (edit: nope, non-json biomes are always kept) When there is Json biomes, the json biome is remade every time the RegistryOps fires.

However, when there are json biomes present, the non-json biome that are not overridden gets their instances passed down through all 2-3 RegistryOps firing and only then. Thus, duplicating the biome modifications to those biomes.

The solution was keeping track of the biomes. But not with a set. Instead, it needs to be a list of biome instances so we know for sure the biome was processed once and only once. 

Now there is some good new, even the non-json biome instances are recreated when exiting and re-entering a world so we need to hook somewhere to know when we exit a world or finished the 2 or 3 RegistryOps firing in order to clear the modified biome list. Any ideas?

**Comment by TelepathicGrunt** [comment]
Ok now it should be good. Just swapping the set of registrykeys to biome instances was enough to solve this and it was already being cleared out when re-entering worlds so there is no memory leak from the new biomes every time a person exits and re-enters worlds without quitting mc. 

Do note, while the test mod has the biome json in its data folder, it seems test mods data folders aren't loaded? But I included the datapack itself so you can just drag and drop it into the datapack screen upon world creation to do the test. The json overridden desert is flatter and has pink skies to show it is the json biome.

**Review by modmuss50 — APPROVED** [review]
Looks good to me 👍 

The build is failing due to checkstyle failiures.

**Review by Technici4n — APPROVED** [review]
Looks good, I will let someone with more biome/registry experience confirm that this is the correct fix.

Once this is merged, a subsequent PR can be made to stabilize this API I believe.

**Review by LambdAurora — APPROVED** [review]
Perfect!

Though I would recommend to not include the custom biome json in the mod resource pack directly has there's no way to disable mod data packs, I would recommend to use a built-in datapack instead in the test module to allow testing without the pack.

**Review by Juuxel — APPROVED** [review]

**Comment by TelepathicGrunt** [comment]
> Perfect!
> 
> Though I would recommend to not include the custom biome json in the mod resource pack directly has there's no way to disable mod data packs, I would recommend to use a built-in datapack instead in the test module to allow testing without the pack.

What do you mean by a built-in datapack? My understanding is either the datapack json is in the mod's data folder which then is read as its own built-in datapack to the game. Or you manually drag the datapack into the datapack screen.

**Comment by Juuxel** [comment]
You can use fabric-resource-loader-v0's `ResourceManagerHelper.registerBuiltinResourcePack` to add builtin but disabled-by-default mod data packs.

**Comment by TelepathicGrunt** [comment]
I can't get the test mod to run if I attempt to use ResourceManagerHelper.registerBuiltinResourcePack as if it cannot ever resolve the imports even when the dependency is added. So instead, I just completely removed the test mod's data folder entirely. The zipped datapack itself is in the resource folder anyway and can just be drag and dropped into a world for the test. I am not gonna let this important PR be hanged up over whatever weird behavior testmods are doing with dependencies and internal data folders. If someone else wants to take up the pain of getting the testmod's data folder to work the way others want it to work, they can amend this PR. Otherwise, drag and dropping in the datapack itself is fine.

**Comment by liach** [comment]
@modmuss50 mind approve running github actions for this?

**Review by liach — APPROVED** [review]

**Review by Player3324 — APPROVED** [review]
This Biome Set won't leak, right?

**Comment by modmuss50** [comment]
I don’t think so as it will go out of scope with the dynamic registry.

I think a better solution to guarantee that will be to add a Boolean into Biome it’s self, but I don’t believe this is an issue.

**Comment by liach** [comment]
can't we always use a resource reloader to clear the set after each data pack reload?

**Comment by shartte** [comment]
```
        		// Vanilla will sometimes call RegistryOps.of twice for the same dynamic registry manager,    
        		// which usually will not result in a reload of objects, since it reuses existing objects
```
(if I remember correctly) refers to some edge cases in vanilla when you create a world, then re-create the world, it would call it twice for the same registry manager. It only applies to singleplayer worlds though. 
Regardless, you should modify the comment though, since now it doesn't match the actual code.

**Pull request merged**